infra/Pos/Communication/Relay/Types.hs

```diff
@@ -44,10 +45,17 @@ instance Buildable PropagationMsg where
     build (DataOnlyPM conts) =
         Buildable.build conts
 
--- | Queue of InvMsges which should be propagated.
-type RelayPropagationQueue = TBQueue (Microsecond, PropagationMsg)
+data RelayContext m = RelayContext
```

Comment:
We want to encapsulate the relay resources in some abstraction `RelayContext`, but since the relayer abstraction itself abstracts over the application monad, we'll have to put it into that type for `RelayContext m`. The application monad itself has access to the `RelayContext m`, and so self-reference in the application monad naturally arises.